backend/src/main/java/com/bakdata/conquery/util/RoundRobinQueue.java

```diff
@@ -0,0 +1,443 @@
+     */
+    private final ThreadLocal<Integer> cycleIndex = ThreadLocal.withInitial(() -> 0);
+
+    public RoundRobinQueue(int capacity) {
+        super();
+        queues = new Queue[capacity];
+    }
+
+	/**
+	 * @return Maximum number of queues allowed in this queue.
+	 */
+	public int getCapacity() {
+		return queues.length;
+	}
+
+	/**
+	 * @return Number of sub-queues.
+	 */
+	public int getNQueues() {
+		return (int) Arrays.stream(queues).filter(Objects::nonNull).count();
+	}
+
+
+	/**
+     * Helper class that notifies on {@code signal} when a new object is added to it queue, awakening waiting threads. This effectively implements a semaphore aroud {@code signal} as notify only awakens a single waiting thread.
+     */
+    @RequiredArgsConstructor
+    private static class SignallingForwardingQueue<T> extends ForwardingQueue<T> {
+
+		/**
+		 * The original queue.
+		 */
+		@NonNull
+        private final Queue<T> base;
+
+		/**
+		 * The signal object to be notified on.
+		 */
+        private final Object signal;
+
+        @Override
+        protected Queue<T> delegate() {
+            return base;
+        }
+
+		protected void doNotify() {
+			log.trace("Awakening a thread for new Work.");
+			synchronized (signal) {
+				signal.notify();
+			}
+		}
+
+		/**
+		 * Try to offer a new element to the queue, if successful notify on signal, awakening a waiting thread.
+		 */
+		@Override
+        public boolean offer(T element) {
+            final boolean offer = super.offer(element);
+
+            if (offer) {
+				doNotify();
+			}
+
+            return offer;
+        }
+
+		/**
+		 * Try to add a new element to the queue, if successful notify on signal, awakening a waiting thread.
+		 */
+        @Override
+        public boolean add(T element) {
+            final boolean add = super.add(element);
+
+            if (add) {
+				doNotify();
+			}
+
+            return add;
+        }
+
+		/**
+		 * Try to offer multiple elements to the queue, if successful notify all threads waiting on signal.
+		 *
+		 * @implNote this can actually cause some threads to receive nothing, as the collections size might be smaller than the number of waiting threads.
+		 */
+		@Override
+		public boolean addAll(Collection<? extends T> collection) {
+			final boolean addAll = super.addAll(collection);
+
+			if(addAll){
+				//TODO does this cause problems?
+				signal.notifyAll();
+			}
+
+			return addAll;
+		}
+	}
+
+	/**
+	 * Create a new queue adding it as sub-queue if there is a slot available.
+	 * @throws IllegalStateException when no more slots are available.
+	 * @return
+	 */
+    public Queue<E> createQueue() {
+        // TODO: 25.06.2020 FK: add supplier as creation parameter
+        final Queue<E> out = new SignallingForwardingQueue<E>(Queues.newConcurrentLinkedQueue(), signal);
+		final int free;
+
+        synchronized (queues) {
+            free = ArrayUtils.indexOf(queues, null);
```

Comment:
Du kannst in dieser Methode das Array auch neu anlegen und vergrößern. Analog in der remove Methode verkleinern. Dann musst du dich nicht mit der festen Kapa rumschlagen.
Oder eine ArrayList. Solange get(i)=> O(1) hat sollte es doch kein Problem sein.

Reply:
Wenn ich das array ändere ist das aber eine racecondition in poll und dann ist das nicht mehr lock-free im consumer